pkg/kademlia/mock/kademlia.go

```diff
@@ -0,0 +1,175 @@
+		for _, a := range addrs {
+			a := a
+			m.eachPeerRev = append(m.eachPeerRev, a)
+		}
+	})
+}
+
+func WithDepth(d uint8) Option {
+	return optionFunc(func(m *Mock) {
+		m.depth = d
+	})
+}
+
+func WithDepthCalls(d ...uint8) Option {
+	return optionFunc(func(m *Mock) {
+		m.depthReplies = d
+	})
+}
+
+type Mock struct {
+	mtx          sync.Mutex
+	peers        []swarm.Address
+	eachPeerRev  []AddrTuple
+	depth        uint8
+	depthReplies []uint8
+	depthCalls   int
+	trigs        []chan struct{}
+	trigMtx      sync.Mutex
+}
+
+func NewMockKademlia(o ...Option) *Mock {
+	m := &Mock{}
+	for _, v := range o {
+		v.apply(m)
+	}
+
+	return m
+}
+
+// AddPeer is called when a peer is added to the topology backlog
+// for further processing by connectivity strategy.
+func (m *Mock) AddPeer(ctx context.Context, addr swarm.Address) error {
+	panic("not implemented") // TODO: Implement
+}
+
+func (m *Mock) ClosestPeer(addr swarm.Address) (peerAddr swarm.Address, err error) {
+	panic("not implemented") // TODO: Implement
+}
+
+// EachPeer iterates from closest bin to farthest
+func (m *Mock) EachPeer(f topology.EachPeerFunc) error {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+
+	for i := len(m.peers) - 1; i > 0; i-- {
+		stop, _, err := f(m.peers[i], uint8(i))
+		if stop {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// EachPeerRev iterates from farthest bin to closest
+func (m *Mock) EachPeerRev(f topology.EachPeerFunc) error {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	for _, v := range m.eachPeerRev {
+		stop, _, err := f(v.A, v.P)
+		if stop {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (m *Mock) NeighborhoodDepth() uint8 {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+
+	m.depthCalls++
+	if len(m.depthReplies) > 0 {
+		return m.depthReplies[m.depthCalls]
+	}
+	return m.depth
+}
+
+// Connected is called when a peer dials in.
+func (m *Mock) Connected(_ context.Context, addr swarm.Address) error {
+	m.mtx.Lock()
+	m.peers = append(m.peers, addr)
+	m.mtx.Unlock()
+	m.Trigger()
+	return nil
+}
+
+// Disconnected is called when a peer disconnects.
+func (m *Mock) Disconnected(_ swarm.Address) {
+	m.Trigger()
+}
+
+func (m *Mock) SubscribePeersChange() (c <-chan struct{}, unsubscribe func()) {
+	channel := make(chan struct{}, 1)
+	var closeOnce sync.Once
+
+	m.trigMtx.Lock()
+	defer m.trigMtx.Unlock()
+	m.trigs = append(m.trigs, channel)
+
+	unsubscribe = func() {
+		m.trigMtx.Lock()
+		defer m.trigMtx.Unlock()
+
+		for i, c := range m.trigs {
+			if c == channel {
+				m.trigs = append(m.trigs[:i], m.trigs[i+1:]...)
+				break
+			}
+		}
+
+		closeOnce.Do(func() { close(channel) })
+	}
+
+	return channel, unsubscribe
+}
+
+func (m *Mock) Trigger() {
+	m.trigMtx.Lock()
+	defer m.trigMtx.Unlock()
+
+	for _, c := range m.trigs {
+		select {
+		case c <- struct{}{}:
+		default:
+		}
+	}
+}
+
+func (m *Mock) Close() error {
+	panic("not implemented") // TODO: Implement
```

Comment:
I think that comment is not needed here as the string in panic is self-desribing

Reply:
this is auto generated text from `:GoImpl` (`vim-go`)